src/main/java/tech/gruppone/stalker/server/services/UserService.java

```diff
@@ -27,27 +31,61 @@
   UserRepository userRepository;
   UserDataRepository userDataRepository;
 
+  private UserDto fromDaosTuple(final Tuple2<UserDao, UserDataDao> tuple) {
+    final var userDao = tuple.getT1();
+    final var userDataDao = tuple.getT2();
+
+    final var userDataDto =
+        UserDataDto.builder()
+            .email(userDao.getEmail())
+            .firstName(userDataDao.getFirstName())
+            .lastName(userDataDao.getLastName())
+            .birthDate(userDataDao.getBirthDate())
+            .creationDateTime(Timestamp.valueOf(userDataDao.getCreatedDate()))
+            .build();
+
+    return new UserDto(userDao.getId(), userDataDto);
+  }
+
+  private Flux<Tuple2<UserDao, UserDataDao>> mergeUsersWithTheirUserData(
+      final Flux<UserDao> userDaos, final Flux<UserDataDao> userDataDaos) {
+
+    return userDaos
+        // this could be refactored to an utility function
+        .flatMap(
+        // zip an userDao with the userDataDao that has the same id
+        userDao -> {
+          final Mono<UserDataDao> matchedUserDataDao =
+              userDataDaos
+                  .filter(userDataDao -> userDataDao.getUserId().equals(userDao.getId()))
+                  .doOnNext(
+                      found -> log.info("found matching userDataDao with id {}", found.getUserId()))
+                  .next();
+
+          return Mono.just(userDao).zipWith(matchedUserDataDao);
+        });
+  }
+
+  public Flux<UserDto> findAll() {
+    final var userDaos = userRepository.findAll();
+    final var userDataDaos = userDataRepository.findAll();
+
+    return mergeUsersWithTheirUserData(userDaos, userDataDaos).map(this::fromDaosTuple);
+  }
+
+  public Flux<UserDto> findAllById(final Flux<Long> ids) {
+    final var userDaos = userRepository.findAllById(ids);
+    final var userDataDaos = userDataRepository.findAllById(ids);
+
+    return mergeUsersWithTheirUserData(userDaos, userDataDaos).map(this::fromDaosTuple);
+  }
+
```

Comment:
Faccio onestamente fatica a capire tutto questo

Reply:
ho tenuto due metodi separati per dare a ciascuno di essi un ruolo specifico; a posteriori non ha così tanto senso. il punto delle `tuple2` è che sono un modo per passare "più di un oggetto" dentro allo stesso mono/flux.

la prima funzione prende in pasto due flux che contengono rispettivamente una lista di UserDao e una lista di UserDataDao (che otteniamo coi metodi di default dei repository) e li "combina" in maniera furba, cioè controllando che lo userdao e lo userdatadao  abbiano lo stesso id. questo è necessario perché r2dbc non garantisce l'ordine dei risultati. inoltre, nel modo in cui l'ho implementato non c'è un impatto irragionevole sulla performance perché il filter che faccio è equivalente a dirgli

"genera un nuovo flux (F2) a partire da quello con tutti gli userdatadao che emette SOLO lo userdatadao con l'id corrispondente, trasformalo in un mono e zippalo con lo userdao"

e tutto questo viene fatto in maniera asincrona (perché siamo dentro a un flatmap), cioè il controllo torna immediatamente all'invoker. quindi se lo userdatadao "in cima al flux" è corrispondente allo userdao "in cima al flux", F2 emette immediatamente, altrimenti rimane la "in attesa" fino a quando dal flux di userdatadao non esce il dao corrispondente